independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java

```diff
@@ -0,0 +1,85 @@
+package io.quarkus.qute;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Escapes a characted sequence using a map of replacements.
+ */
+public final class Escaper {
+
+    private final Map<Character, String> replacements;
+
+    /**
+     *
+     * @param replacements
+     */
+    private Escaper(Map<Character, String> replacements) {
+        this.replacements = replacements.isEmpty() ? Collections.emptyMap()
+                : new HashMap<>(replacements);
+    }
+
+    /**
+     *
+     * @param value
+     * @return an escaped value
+     */
+    public String escape(CharSequence value) {
+        Objects.requireNonNull(value);
+        if (value.length() == 0) {
+            return value.toString();
+        }
+        for (int i = 0; i < value.length(); i++) {
+            String replacement = replacements.get(value.charAt(i));
```

Comment:
The auto-boxing here looks like an expensive operation to be done on every character.

Reply:
Well, `Map#get()` is probably even more expensive. Anyway, it's an implementation detail that can be improved in the future... feel free to improve the algorithm ;-).

Reply:
Yeah, I was looking at https://www.eclipse.org/collections/javadoc/10.0.0/org/eclipse/collections/impl/map/mutable/primitive/CharObjectHashMap.html. I'll wait until your PR is merged to perform some performance tests using it ;)

Reply:
There is also `io.netty.util.collection.CharObjectHashMap<V>`.

Reply:
Yeah, I've been looking at that too initially. Maybe it's simple to just copy it over to Qute than to depend on another library. OSS FTW :)